src/libraries/System.IO.Compression/src/System/IO/Compression/ZipCryptoStream.cs

```diff
@@ -0,0 +1,254 @@
+
+            Span<byte> hdrPlain = stackalloc byte[12];
+
+            // bytes 0..9 are random
+            // TODO: change to actual random data later
+            for (int i = 0; i < 10; i++)
+                hdrPlain[i] = 0;
+
+            // bytes 10..11: check bytes (CRC-based if crc32 provided; else DOS time low word)
+            if (_crc32ForHeader.HasValue)
+            {
+                uint crc = _crc32ForHeader.Value;
+                hdrPlain[10] = (byte)((crc >> 16) & 0xFF);
+                hdrPlain[11] = (byte)((crc >> 24) & 0xFF);
+            }
+            else
+            {
+                hdrPlain[10] = (byte)(_verifierLow2Bytes & 0xFF);
+                hdrPlain[11] = (byte)((_verifierLow2Bytes >> 8) & 0xFF);
+            }
+
+            // Encrypt & write; update keys with PLAINTEXT per spec
+            byte[] hdrCiph = new byte[12];
+            for (int i = 0; i < 12; i++)
+            {
+                byte ks = DecipherByte();
+                byte p = hdrPlain[i];
+                hdrCiph[i] = (byte)(p ^ ks);
+                UpdateKeys(p);
+            }
+
+            _base.Write(hdrCiph, 0, 12);
+            _headerWritten = true;
+            _position += 12;
+        }
+
+        private void InitKeysFromBytes(ReadOnlySpan<char> password)
+        {
+            _key0 = 305419896;
+            _key1 = 591751049;
+            _key2 = 878082192;
+
+            // ZipCrypto uses raw bytes; ASCII is the most interoperable (UTF8 also acceptable).
+            var bytes = password.ToArray();
+            foreach (byte b in bytes)
+                UpdateKeys(b);
+        }
+
+        private void ValidateHeader(byte expectedCheckByte)
+        {
+            byte[] hdr = new byte[12];
+            int read = 0;
+            while (read < hdr.Length)
+            {
+                int n = _base.Read(hdr, read, hdr.Length - read);
+                if (n <= 0) throw new InvalidDataException("Truncated ZipCrypto header.");
+                read += n;
+            }
+
+            for (int i = 0; i < hdr.Length; i++)
+                hdr[i] = DecryptByte(hdr[i]);
+
+            if (hdr[11] != expectedCheckByte)
+                throw new InvalidDataException("Invalid password for encrypted ZIP entry.");
+        }
+
+        private void UpdateKeys(byte b)
+        {
+            _key0 = Crc32Update(_key0, b);
+            _key1 += (_key0 & 0xFF);
+            _key1 = _key1 * 134775813 + 1;
+            _key2 = Crc32Update(_key2, (byte)(_key1 >> 24));
+        }
+
+        private byte DecipherByte()
+        {
+            uint temp = _key2 | 2; // use uint to avoid narrowing issues
+            return (byte)((temp * (temp ^ 1)) >> 8);
+        }
+
+        private byte DecryptByte(byte ciph)
+        {
+            byte m = DecipherByte();
+            byte plain = (byte)(ciph ^ m);
+            UpdateKeys(plain);
+            return plain;
+        }
+
+        // ---- Stream overrides ----
```

Comment:
This will also need WriteAsync and ReadAsync overloads. And since you are doing reads in constructor, then some way to do that part async as well.